onnxruntime/core/providers/coreml/builders/impl/split_op_builder.cc

```diff
@@ -0,0 +1,179 @@
+                                             const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  std::vector<int64_t> data_shape;
+  ORT_RETURN_IF_NOT(GetStaticShape(*node.InputDefs()[0], data_shape, logger), "Failed to get input shape.");
+
+  NodeAttrHelper helper(node);
+  const auto axis = helper.Get("axis", 0);
+
+  // attribute introduced since opset 18
+  uint64_t num_outputs;
+
+  std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
+  auto* coreml_splitnd = layer->mutable_splitnd();
+  coreml_splitnd->set_axis(axis);
+
+  if (input_defs.size() > 1) {
+    // if "split" is explicitly provided as an input
+    const auto& split_tensor = *model_builder.GetInitializerTensors().at(input_defs[1]->Name());
+    Initializer unpacked_tensor(split_tensor);
+    auto split_span = unpacked_tensor.DataAsSpan<int64_t>();
+    auto split_sizes = split_span.size();
+    num_outputs = SafeInt<uint64_t>(split_sizes);
+    for (size_t i = 0; i < split_sizes; i++) {
+      coreml_splitnd->add_splitsizes(SafeInt<uint64_t>(split_span[i]));
+    }
+  } else if (node.SinceVersion() < 18) {
+    num_outputs = node.OutputDefs().size();
+    coreml_splitnd->set_numsplits(num_outputs);
+  } else {
+    num_outputs = SafeInt<uint64_t>(helper.Get("num_outputs", 2));
+    auto split_dim_size = data_shape[HandleNegativeAxis(axis, data_shape.size())];
+    uint64_t chunk_size = SafeInt<uint64_t>((split_dim_size + num_outputs - 1) / num_outputs);
+    uint64_t remainder = split_dim_size % chunk_size;
+    if (remainder) {
+      // uneven
+      auto split_sizes = std::vector<uint64_t>(num_outputs, chunk_size);
+      split_sizes.back() = remainder;
+      for (size_t i = 0; i < split_sizes.size(); i++) {
+        coreml_splitnd->add_splitsizes(SafeInt<uint64_t>(split_sizes[i]));
+      }
+    } else {
+      // even
+      coreml_splitnd->set_numsplits(num_outputs);
+    }
+  }
+
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  // variadic number of outputs. Calculated based on the length of the given splitSizes if provided.
+  // Otherwise, uses attribute value 'num_outputs'.
+  for (uint64_t i = 0; i < num_outputs; i++) {
+    *layer->mutable_output()->Add() = node.OutputDefs()[i]->Name();
+  }
+  model_builder.AddLayer(std::move(layer));
+
+  return Status::OK();
+}
+
+#endif
+
+// Operator support related
+
+bool SplitOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                                       const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers = input_params.graph_viewer.GetAllInitializedTensors();
+
+  NodeAttrHelper helper(node);
+  const auto axis = helper.Get("axis", 0);
+  const auto num_outputs = helper.Get("num_outputs", -1);
+
+  std::vector<int64_t> input_shape;
+  if (!GetStaticShape(*input_defs[0], input_shape, logger))
+    return false;
```

Comment:
Does the entire shape need to be static or just the dimension we're splitting on? Wondering if this is too strict for the new dynamic input shape support in CoreML.

Reply:
I am actually not so clear about the criteria that when to call `GetStaticShape` when to call `GetShape` in CoreML.

I updated for split op here to use GetShape() for now.

Reply:
> I am actually not so clear about the criteria between when to call GetStaticShape vs. when to call GetShape for different ops in CoreML.

if you need a shape with no dynamic dimensions, call `GetStaticShape`. otherwise, if you just need a shape and it's fine for it to have dynamic (-1) dimensions, call `GetShape`.